### PR TITLE
Reject concurrent transactions

### LINKED ISSUE #9489: Incoming federation transactions stack up indefinitely
Since #8342, we apply the federation ratelimiter *after* the transaction linearizer. The net effect is that, if we cannot keep up with incoming transactions, we allow them to stack up indefinitely (there can only be one thread inside the transaction linearizer, so we never hit the concurrency limit on the ratelimiter).  This can use a lot of memory, and it's not the way matrix federation transactions are meant to work (if the remote gets behind, it should drop messages until it catches up).

It feels like the ratelimter is pretty redundant there. (Admittedly it still has some use as a true *rate* limiter rather than a concurrency limiter, but that's a bit of an edge-case.)

Should we instead be using a separate ratelimiter dedicated to incoming transactions?

### PR DESCRIPTION
**Pull request by richvdh**
If more transactions arrive from an origin while we're still processing the first one, reject them.

Hopefully fixes #9489

### CODE REVIEW

File: synapse/federation/federation_server.py [hunk 1]
```diff
@@ -169,6 +170,33 @@ async def on_incoming_transaction(
 
         logger.debug("[%s] Got transaction", transaction_id)
 
+        # Reject malformed transactions early: reject if too many PDUs/EDUs
+        if len(transaction.pdus) > 50 or (  # type: ignore
+            hasattr(transaction, "edus") and len(transaction.edus) > 100  # type: ignore
+        ):
+            logger.info("Transaction PDU or EDU count too large. Returning 400")
+            return 400, {}
```
Comment (richvdh) [inline synapse/federation/federation_server.py:178]: I moved this up here from `_handle_incoming_transaction` to keep a sytest (which was sending concurrent transactions) happy, but it seems like generally a good thing to do.

File: synapse/federation/federation_server.py [hunk 2]
```diff
@@ -112,10 +112,11 @@ def __init__(self, hs: "HomeServer"):
         # with FederationHandlerRegistry.
         hs.get_directory_handler()
 
-        self._federation_ratelimiter = hs.get_federation_ratelimiter()
```
Comment (clokep) [inline synapse/federation/federation_server.py:115]: I wonder if we should just inline the rate-limiter that used to be shared with this into `TransportLayerServer` now...

**Review by clokep — APPROVED** [review]
Seems to make sense. It definitely makes the code easier to reason about!

Synapse has always retried federation transactions on a 429, correct? This won't break old servers federating with newer servers?

File: synapse/federation/federation_server.py (hunk 1 above)
Comment (clokep) [reply synapse/federation/federation_server.py:178]: I agree that it seems best to do this as early as possible. 👍

**Comment by richvdh** [comment]
> Synapse has always retried federation transactions on a 429, correct? 

looks like it might have changed in #4358, but I don't think it's going to break anything too badly anyway. Worst case the sender will back off transmission for a bit, which is fine.

File: synapse/federation/federation_server.py (hunk 2 above)
Comment (richvdh) [reply synapse/federation/federation_server.py:115]: :man_shrugging: probably...

**Comment by richvdh** [comment]
> looks like it might have changed in #4358

actually looks like we retried for all response codes before that.

**Pull request merged**